include/cpu/mos_6502.h

```diff
@@ -0,0 +1,345 @@
+// OsNES Project | include/mos_6502.h
+// Description:
+// Functional specification for the MOS 6502 CPU.
+// Any and all details related to the overall operation ofthe 6502 should
+// be declared here.
+//
+// Authors: 
+// John Dengis
+//
+#ifndef MOS_6502_H
+#define MOS_6502_H
+#include "common.h"
+#include "cpu.h"
+#include "mem.h"
+
+class MOS_6502 : public CPU<byte> {
+  public:
+    void  execute(byte opcode) override;
+    byte fetch() override;
+
+  protected:
+    // Address mode functions
+    Mem::Ref accumulator();
+    Mem::Ref absolute();
+    Mem::Ref absolute_X();
+    Mem::Ref absolute_Y();
+    Mem::Ref immediate();
+    Mem::Ref indirect();
+
+    // CPU Instruction emulation functions
+    void ADC(const Mem::Ref);
+    void AND(const Mem::Ref);
+    void ASL(Mem::Ref);
+    void BCC(const Mem::Ref);
+    void BCS(const Mem::Ref);
+    void BEQ(const Mem::Ref);
+    void BIT(const Mem::Ref);
+    void BMI(const Mem::Ref);
+    void BNE(const Mem::Ref);
+    void BPL(const Mem::Ref);
+    void BRK();
+    void BVC(const Mem::Ref);
+    void BVS(const Mem::Ref);
+    void CLC();
+    void CLD();
+    void CLI();
+    void CLV();
+    void CMP(const Mem::Ref);
+    void CPX(const Mem::Ref);
+    void CPY(const Mem::Ref);
+    void DEC(Mem::Ref);
+    void DEX();
+    void DEY();
+    void EOR(const Mem::Ref);
+    void INC(Mem::Ref);
+    void INX();
+    void INY();
+    void JMP(const Mem::Ref);
+    void JSR(const Mem::Ref);
+    void LDA(const Mem::Ref);
+    void LDX(const Mem::Ref);
+    void LDY(const Mem::Ref);
+    void LSR(Mem::Ref);
+    void NOP();
+    void ORA(const Mem::Ref);
+    void PHA();
+    void PHP();
+    void PLA();
+    void PLP();
+    void ROL(Mem::Ref);
+    void ROR(Mem::Ref);
+    void RTI();
+    void RTS();
+    void SBC(const Mem::Ref);
+    void SEC();
+    void SED();
+    void SEI();
+    void STA(Mem::Ref);
+    void STX(Mem::Ref);
+    void STY(Mem::Ref);
+    void TAX();
+    void TAY();
+    void TSX();
+    void TXA();
+    void TXS();
+    void TYA();
+
+  private:
+    void ADC_Impl(const byte);
```

Comment:
You shouldn't need this you baddie. Just do this thing: Use pass-by-value to send arguments to the functions, and use a return value to send back the data that must be added to memory in the event that memory must be altered (use of the reference to memory will happen in Execute())